src/Autofac/Util/AssemblyExtensions.cs

```diff
@@ -39,4 +40,21 @@ public static IEnumerable<Type> GetLoadableTypes(this Assembly assembly)
             return ex.Types.Where(t => t is not null)!;
         }
     }
+
+    /// <summary>
+    /// Get the set of types that Autofac will allow to be loaded from the given assembly.
+    /// </summary>
+    /// <param name="assembly">The assembly to load types from.</param>
+    /// <returns>The set of loadable types.</returns>
+    internal static IEnumerable<Type> GetPermittedTypesForAssemblyScanning(this Assembly assembly)
+    {
+        static IReadOnlyList<Type> Uncached(Assembly assembly)
+        {
+            return assembly.GetLoadableTypes()
+                           .Where(t => t.IsClass && !t.IsAbstract && !t.IsDelegate() && !t.IsCompilerGenerated())
```

Comment:
There was a comment in `OpenGenericScanningRegistrationExtensions.cs` about why `IsCompilerGenerated` is intentionally last here. Might be good to keep that; I'd never have realized there's a perf issue attached to it.